tests/runtime/engine/runner.py

```diff
@@ -196,42 +232,64 @@ def run_test(test):
                     elif bpffeature[feature]:
                         print(warn("[   SKIP   ] ") + "%s.%s" % (test.suite, test.name))
                         return Runner.SKIP_FEATURE_REQUIREMENT_UNSATISFIED
+            
+            if test.new_pidns and not test.befores:
+                raise ValueError("`NEW_PIDNS` requires at least one `BEFORE` directive as something needs to run in the new pid namespace")
 
             if test.befores:
-                for before in test.befores:
-                    before = subprocess.Popen(before.split(),
-                                              start_new_session=True,
-                                              universal_newlines=True,
-                                              stdout=subprocess.PIPE,
-                                              stderr=subprocess.STDOUT)
-                    befores.append(before)
-
-                with open(os.devnull, 'w') as dn:
+                if test.new_pidns:
+                    # Use the first BEFORE as the first process in the new pid namespace
+                    unshare_out = subprocess.Popen(["unshare", "--fork", "--pid", "--mount-proc", "-r", "--kill-child"] + test.befores[0].split(),
+                                                   start_new_session=True, universal_newlines=True,
+                                                   stdout=subprocess.PIPE,
+                                                   stderr=subprocess.STDOUT)
+
+                    def has_children(lines):
+                        if len(lines) == 0:
+                            return False
+                        nsenter.extend(["nsenter", "-p", "-m", "-t", lines[0].strip()])
```

Comment:
A bit confusing that the "conditional" callback mutates state in the caller

Reply:
I know. I didn't love this either. Maybe I could have this function return the lines if the conditional is true.